### ISSUE TITLE
Writing this in Java help

### ISSUE THREAD

**Issue by LCCoolJ95** [smogon/pokemon-showdown#6357]
Hey, so I wanna create a Pokémon Battle System in Java using this code. However, it's kind of strange since I can't figure out how the actual functions and events go. It's all different, and the data, or I guess the constants are very different from JavaScript to Java. For example, I get events in JavaScript are like functions/methods in Java, but I don't know where they are. And some .js files use something at the beginning of the file, 'use strict'. I don't have an idea of what that means. Is there some decent documentation, or help, where all the events are, or some idea of how the battle system works? How abilities are called, who moves first, how items are called during battle, etc.

**Comment by xfix** [comment]
You may find https://github.com/smogon/pokemon-showdown/blob/master/simulator-doc.txt useful.

**Comment by Volectric** [comment]
Funnily enough, I am trying to do something similar as well. 
A lot of your problems come with the differences and features between Java and JavaScript. For example, Look at the following JavaScript code:

```
var obj = {
    data: 6,
    moreData: 3
};
console.log(obj.data + " " + obj.moreData);
// prints "6 3"
console.log("Array brackets with strings are functionally equivalent: " + obj["data"] + " " + obj["moreData"]);
// prints "Array brackets with strings are functionally equivalent: 6 3"
```

So, for the pokedex file, you could access Bulbasaur's types by using "BattlePokedex.bulbasaur.types" or "BattlePokedex['bulbasaur']['types']" instead, which is nice since it lets you use a string variable with pokemon's name instead of an int with an ID number. However, you can't really do that exact thing (or at least, not easily or very efficiently) in Java.

JavaScript also features classless objects, which don't exist in Java. Essentially, if you want to do this in Java, you have to use Java features in place of the JavaScript features you are using, which should be expected.

Options for items (as you said you had trouble with) could either be an array of an item class that contains all the item details and a function (or functions, because items activate at different times) that passes the Item's name as a parameter to check if the item does anything as needed; or, you could make an abstract class called Item, with subclasses for each item, with their own abstract function definitions, which are called in the body of the program if the pokemon has the item.

**TL:DR** You just have to get creative with what you have. I still believe I'm pretty novice, so I'm sorry if I got anything wrong or didn't help very much.

Hope I Helped!

**Comment by Zarel** [comment]
I'm confused. Are you trying to port this codebase to Java?

That sounds like it would be very hard, yes. A lot of things that are objects in PS would probably be best represented as dictionaries in Java.

Events in JavaScript are unrelated to methods in Java. JavaScript has methods, too; they mostly work the same way as Java methods. Events are really just a pattern for running some code when something else happens. For us, it lets us organize the "do something when a pokemon with Rocky Helmet gets hit" code in the Rocky Helmet file, instead of in the pokemon-gets-hit file.

The battle system is pretty poorly documented right now, but even if it was well-documented, this is really not a good project for a novice. It took multiple developers years to write this codebase; understanding and porting it is just not something anyone should be doing casually.

**Comment by Zarel** [comment]
Anyway, this issue tracker is not actually a tech support desk. None of what you're asking is a problem in PS itself.
